core/src/banking_stage/transaction_scheduler/scheduler_controller.rs

```diff
@@ -213,8 +232,113 @@ impl SchedulerController {
+        const CHUNK_SIZE: usize = 64;
+        let mut num_forwarded: usize = 0;
+        let mut ids_to_add_back = Vec::new();
+        let mut max_time_reached = false;
+        while !self.container.is_empty() {
+            let mut filter_array = [true; CHUNK_SIZE];
+            let mut ids = Vec::with_capacity(CHUNK_SIZE);
+            let mut txs = Vec::with_capacity(CHUNK_SIZE);
+
+            for _ in 0..CHUNK_SIZE {
+                if let Some(id) = self.container.pop() {
+                    ids.push(id);
+                } else {
+                    break;
+                }
+            }
+            let chunk_size = ids.len();
+            ids.iter().for_each(|id| {
+                let transaction = self.container.get_transaction_ttl(&id.id).unwrap();
+                txs.push(&transaction.transaction);
+            });
+
+            // use same filter we use for processing transactions:
+            // age, already processed, fee-check.
+            Self::pre_graph_filter(
+                &txs,
+                &mut filter_array,
+                &bank,
+                MAX_PROCESSING_AGE
+                    .saturating_sub(FORWARD_TRANSACTIONS_TO_LEADER_AT_SLOT_OFFSET as usize),
+            );
+
+            for (id, filter_result) in ids.iter().zip(&filter_array[..chunk_size]) {
+                if !*filter_result {
+                    self.container.remove_by_id(&id.id);
+                    continue;
+                }
+
+                ids_to_add_back.push(*id); // add back to the queue at end
+                let state = self.container.get_mut_transaction_state(&id.id).unwrap();
+                let sanitized_transaction = &state.transaction_ttl().transaction;
+                let immutable_packet = state.packet().clone();
+
+                // If not already forwarded and can be forwarded, add to forwardable packets.
+                if state.should_forward()
+                    && forwardable_packets.try_add_packet(
+                        sanitized_transaction,
+                        immutable_packet,
+                        feature_set,
+                    )
+                {
+                    saturating_add_assign!(num_forwarded, 1);
+                    state.mark_forwarded();
+                }
+            }
+
+            if start.elapsed() >= MAX_FORWARDING_DURATION {
+                max_time_reached = true;
+                break;
+            }
+        }
+
+        // Push into worker queue.
+        for batch in forwardable_packets.take_iter_batches() {
+            let packets = batch.take_packets();
+            let forward_work = ForwardWork { packets };
+            self.forward_work_sender.send(forward_work).unwrap();
+        }
+
+        // If we hit the time limit. Drop everything that was not checked/processed.
+        // If we cannot run these simple checks in time, then we cannot run them during
+        // leader slot.
+        if max_time_reached {
```

Comment:
how's that 100 ms determined?

Reply:
Just did 1/4 of a block time; the checks done in this should be really quick, and if we cannot do these within 100ms we won't get to these packets in a scheduling loop which does more complex checks than these (currently).

Reply:
The logic of "we will do up to 100ms forwarding work, if there are still packets in container after that, they will be dropped" sounds a bit arbitrary. When I benching it, I have to change `MAX_FORWARDING_DURATION` to `MAX`. What do you think to stop processing forwarding when CU limit is hit (it already does that somewhat, just not perfect).

In any case, for this PR we worry about sending too many to next leader more than not sending enough. I think it is OK for now. Maybe keep an eye on number of forwarded packets vs dropped packets.